packages/pug-lexer/index.js

```diff
@@ -990,172 +1046,244 @@ Lexer.prototype = {
+  /**
+   * Attributes.
+   */
+  
+  attrs: function() {
+    var tok;
+    
+    if ('(' == this.input.charAt(0)) {
+      tok = this.tok('start-attributes');
+      var index = this.bracketExpression().end;
+      var str = this.input.substr(1, index-1);
 
-          var tok = this.tok('attribute');
-          tok.name = key;
-          tok.val = '' == val ? true : val;
-          tok.col = colnoBeginAttr;
-          tok.mustEscape = escapedAttr;
-          this.tokens.push(tok);
+      this.incrementColumn(1);
+      this.tokens.push(this.tokEnd(tok));
+      this.assertNestingCorrect(str);
+      this.consume(index + 1);
 
-          key = val = '';
-          loc = 'key';
-          escapedAttr = false;
-          this.lineno = lineno;
-        } else {
-          switch (loc) {
-            case 'key-char':
-              if (str[i] === quote) {
-                loc = 'key';
-                if (i + 1 < str.length && !/[ ,!=\n\t]/.test(str[i + 1]))
-                  this.error('INVALID_KEY_CHARACTER', 'Unexpected character "' + str[i + 1] + '" expected ` `, `\\n`, `\t`, `,`, `!` or `=`');
-              } else {
-                key += str[i];
-              }
-              break;
-            case 'key':
-              if (key === '' && quoteRe.test(str[i])) {
-                loc = 'key-char';
-                quote = str[i];
-              } else if (str[i] === '!' || str[i] === '=') {
-                escapedAttr = str[i] !== '!';
-                if (str[i] === '!') {
-                  this.incrementColumn(1);
-                  i++;
-                }
-                if (str[i] !== '=') this.error('INVALID_KEY_CHARACTER', 'Unexpected character ' + str[i] + ' expected `=`');
-                loc = 'value';
-                colnoBeginVal = this.colno + 1;
-                state = characterParser.defaultState();
-              } else {
-                key += str[i]
-              }
-              break;
-            case 'value':
-              state = characterParser.parseChar(str[i], state);
-              val += str[i];
-              break;
-          }
-        }
-        if (str[i] === '\n') {
-          // Save the line number locally to keep this.lineno at the start of
-          // the attribute.
-          lineno++;
-          this.colno = 1;
-          // If the key has not been started, update this.lineno immediately.
-          if (!key.trim()) this.lineno = lineno;
-        } else if (str[i] !== undefined) {
-          this.incrementColumn(1);
-        }
+      while(str){
+        str = this.attributeName(str);
       }
 
-      // Reset the line numbers based on the line started on
-      // plus the number of newline characters encountered
-      this.lineno = startingLine + (str.match(/\n/g) || []).length;
-
-      this.tokens.push(this.tok('end-attributes'));
+      tok = this.tok('end-attributes');
       this.incrementColumn(1);
+      this.tokEnd(tok);
+      this.tokens.push(tok);
```

Comment:
In most places you have done `this.tokens.push(this.tokEnd(tok));` - I think we should be consistent here.